File: lang/tests-js/src/test/scala/com/wavesplatform/JsTestBase.scala
package com.wavesplatform

import com.wavesplatform.lang.directives.values.{StdLibVersion, V6}
import testHelpers.TestDataConstantsAndMethods
import utest.{TestSuite, assert}

import scala.scalajs.js.{Dictionary, isUndefined}

abstract class JsTestBase extends TestSuite {

  protected val testData: TestDataConstantsAndMethods.type = TestDataConstantsAndMethods
  protected def assertCompileError(code: String, expectingError: String, estimator: Int = 3): Unit = {
    val error = JsAPI.compile(code, estimator).error
    assert(error.toString.contains(expectingError))
  }

  protected def assertCompileErrorDApp(code: String, version: StdLibVersion, expectingError: String, estimator: Int = 3): Unit = {
    val error = JsAPI.compile(dApp(code, version), estimator).error
    assert(error.toString.contains(expectingError))
  }

  protected def assertCompileErrorExpression(code: String, version: StdLibVersion, expectingError: String, estimator: Int = 3): Unit = {
    val error = JsAPI.compile(expression(code, version), estimator).error
    assert(error.toString.contains(expectingError))
  }

  protected def assertCompileSuccess(code: String, estimator: Int = 3): Unit = {
    val error = JsAPI.compile(code, estimator).error
    assert(isUndefined(error))
  }

  protected def assertCompileSuccessDApp(code: String, version: StdLibVersion, estimator: Int = 3): Unit = {
    val error = JsAPI.compile(dApp(code, version), estimator).error
    assert(isUndefined(error))
  }

  protected def assertCompileSuccessExpression(code: String, version: StdLibVersion, estimator: Int = 3): Unit = {
    val error = JsAPI.compile(expression(code, version), estimator).error
    assert(isUndefined(error))
  }

  protected def expressionComplexity(code: String, version: StdLibVersion = V6, estimator: Int = 3): Int =
    JsAPI.compile(expression(code, version), estimator).complexity.asInstanceOf[Int]

  protected def dAppComplexities(code: String, version: StdLibVersion = V6, estimator: Int = 3): DAppComplexities = {
    val result = JsAPI.compile(dApp(code, version), estimator)
    DAppComplexities(
      result.complexity.asInstanceOf[Int],
      result.verifierComplexity.asInstanceOf[Int],
      result.callableComplexities.asInstanceOf[Dictionary[Int]].toMap,
      result.userFunctionComplexities.asInstanceOf[Dictionary[Int]].toMap,
      result.globalVariableComplexities.asInstanceOf[Dictionary[Int]].toMap
    )
  }

  protected def expression(code: String, version: StdLibVersion): String =
    s"""
       |{-# STDLIB_VERSION ${version.id} #-}
       |{-# CONTENT_TYPE   EXPRESSION    #-}
       |{-# SCRIPT_TYPE    ACCOUNT       #-}
       |
       |$code
     """.stripMargin

  protected def dApp(code: String, version: StdLibVersion): String =
    s"""
       |{-# STDLIB_VERSION ${version.id} #-}
       |{-# CONTENT_TYPE   DAPP          #-}
       |{-# SCRIPT_TYPE    ACCOUNT       #-}
       |
       |$code
     """.stripMargin
}

File: ride-js-bundle/test/helpers/jsTestBase.ts
// Port of com/wavesplatform/JsTestBase.scala
// Assertion helpers + script wrappers built on the public compiler API.
import { expect } from "vitest";
import { compile } from "./compiler";

// StdLibVersion collapses to its numeric id in JS.
export const V1 = 1;
export const V2 = 2;
export const V3 = 3;
export const V4 = 4;
export const V5 = 5;
export const V6 = 6;
export const V7 = 7;
export const V8 = 8;
export const V9 = 9;

// Mirrors DirectiveDictionary[StdLibVersion].all (all defined library versions).
export const allStdLibVersions: number[] = [V1, V2, V3, V4, V5, V6, V7, V8, V9];

export interface DAppComplexitiesResult {
  complexity: number;
  verifierComplexity: number;
  callableComplexities: Record<string, number>;
  userFunctionComplexities: Record<string, number>;
  globalVariableComplexities: Record<string, number>;
}

export function assertCompileError(code: string, expectingError: string, estimator = 3): void {
  const error = compile(code, estimator).error;
  expect(String(error)).toContain(expectingError);
}

export function assertCompileErrorDApp(code: string, version: number, expectingError: string, estimator = 3): void {
  const error = compile(dApp(code, version), estimator).error;
  expect(String(error)).toContain(expectingError);
}

export function assertCompileErrorExpression(code: string, version: number, expectingError: string, estimator = 3): void {
  const error = compile(expression(code, version), estimator).error;
  expect(String(error)).toContain(expectingError);
}

export function assertCompileSuccess(code: string, estimator = 3): void {
  const error = compile(code, estimator).error;
  expect(error).toBeUndefined();
}

export function assertCompileSuccessDApp(code: string, version: number, estimator = 3): void {
  const error = compile(dApp(code, version), estimator).error;
  expect(error).toBeUndefined();
}

export function assertCompileSuccessExpression(code: string, version: number, estimator = 3): void {
  const error = compile(expression(code, version), estimator).error;
  expect(error).toBeUndefined();
}

export function expressionComplexity(code: string, version = V6, estimator = 3): number {
  // The wrapped public `compile` nests a successful result under `.result`.
  return compile(expression(code, version), estimator).result.complexity as number;
}

export function dAppComplexities(code: string, version = V6, estimator = 3): DAppComplexitiesResult {
  const { result } = compile(dApp(code, version), estimator);
  return {
    complexity: result.complexity as number,
    verifierComplexity: result.verifierComplexity as number,
    callableComplexities: result.callableComplexities as Record<string, number>,
    userFunctionComplexities: result.userFunctionComplexities as Record<string, number>,
    globalVariableComplexities: result.globalVariableComplexities as Record<string, number>
  };
}

export function expression(code: string, version: number): string {
  return `
{-# STDLIB_VERSION ${version} #-}
{-# CONTENT_TYPE   EXPRESSION    #-}
{-# SCRIPT_TYPE    ACCOUNT       #-}

${code}
     `;
}

export function dApp(code: string, version: number): string {
  return `
{-# STDLIB_VERSION ${version} #-}
{-# CONTENT_TYPE   DAPP          #-}
{-# SCRIPT_TYPE    ACCOUNT       #-}

${code}
     `;
}
